src/helper/remote_config/product.hpp

```diff
@@ -23,39 +33,69 @@ class product_listener_base {
     product_listener_base &operator=(product_listener_base &&) = default;
     virtual ~product_listener_base() = default;
 
-    virtual void on_update(const std::map<std::string, config> &configs) = 0;
-    virtual void on_unapply(const std::map<std::string, config> &configs) = 0;
+    virtual void on_update(const config &config) = 0;
+    virtual void on_unapply(const config &config) = 0;
 };
 
 class product {
 public:
-    explicit product(
-        std::string &&name, std::vector<product_listener_base *> &&listeners)
-        : name_(std::move(name)), listeners_(std::move(listeners)){};
+    product(std::string &&name, product_listener_base *listener)
+        : name_(std::move(name)), listener_(listener){};
+
     void assign_configs(const std::map<std::string, config> &configs)
     {
         std::map<std::string, config> to_update;
         std::map<std::string, config> to_keep;
-
         for (const auto &config : configs) {
             auto previous_config = configs_.find(config.first);
             if (previous_config == configs_.end()) { // New config
                 to_update.emplace(config.first, config.second);
             } else { // Already existed
                 if (config.second.hashes ==
                     previous_config->second.hashes) { // No changes in config
-                    to_keep.emplace(config.first, config.second);
+                    to_keep.emplace(config.first, previous_config->second);
                 } else { // Config updated
                     to_update.emplace(config.first, config.second);
                 }
                 configs_.erase(previous_config);
```

Comment:
This is causing configuration to be "unapplied", so some comments here would be useful.

Reply:
Comment still missing...

Reply:
It's my understanding that you're removing elements from `configs_` so at the end you only have left those which need to be unapplied, is that not correct? So it would be equivalent to an "to_unapply" map.